components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,387 @@
+        Ok(file_dict)
+    }
+
+    /// Append an insert operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn insert(&mut self, name: &str, info: FileInfo) -> Result<()> {
+        let file = self.append_file.as_mut().unwrap();
+        let bytes = Self::convert_record_to_bytes(name, LogRecord::INSERT(info.clone()))?;
+        file.write_all(&bytes)?;
+        file.sync_all()?;
+
+        self.file_dict.files.insert(name.to_owned(), info);
+        self.check_compact()?;
+
+        Ok(())
+    }
+
+    /// Append a remove operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn remove(&mut self, name: &str) -> Result<()> {
+        let file = self.append_file.as_mut().unwrap();
+        let bytes = Self::convert_record_to_bytes(name, LogRecord::REMOVE)?;
+        file.write_all(&bytes)?;
+        file.sync_all()?;
+
+        self.file_dict.files.remove(name);
+        self.check_compact()?;
+
+        Ok(())
+    }
+
+    /// Replace existed file entry with new file information.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn replace(&mut self, src_name: &str, dst_name: &str, info: FileInfo) -> Result<()> {
+        assert_ne!(src_name, dst_name);
+        let file = self.append_file.as_mut().unwrap();
+        let bytes1 = Self::convert_record_to_bytes(dst_name, LogRecord::INSERT(info.clone()))?;
+        let bytes2 = Self::convert_record_to_bytes(src_name, LogRecord::REMOVE)?;
+        file.write_all(&bytes1)?;
+        file.write_all(&bytes2)?;
+        file.sync_all()?;
+
+        self.file_dict.files.remove(src_name);
+        self.file_dict.files.insert(dst_name.to_owned(), info);
+        self.check_compact()?;
+
+        Ok(())
+    }
+
+    /// This function needs to be called after each append operation to check
+    /// if compact is needed.
+    fn check_compact(&mut self) -> Result<()> {
+        if self.file_dict.files.len() as u64 > Self::REWRITE_THRESHOLD {
+            self.write(&self.file_dict.clone())?;
+        }
+        Ok(())
+    }
+
+    fn convert_record_to_bytes(fname: &str, record_type: LogRecord) -> Result<Vec<u8>> {
+        let name_len = fname.len() as u16;
+        let mut content = vec![];
+        content.extend_from_slice(fname.as_bytes());
+
+        let mut header_buf = [0; Self::RECORD_HEADER_SIZE];
+        let info_len = match record_type {
+            LogRecord::INSERT(info) => {
+                header_buf[8] = 1;
+                let info_bytes = info.write_to_bytes()?;
+                content.extend_from_slice(&info_bytes);
+                info_bytes.len() as u16
+            }
+            LogRecord::REMOVE => {
+                header_buf[8] = 2;
+                0
+            }
+        };
+
+        let mut digest = crc32fast::Hasher::new();
+        digest.update(&content);
+        let crc32 = digest.finalize();
```

Comment:
let crc should also checksum `name_len` and `info_len`?

Reply:
https://github.com/tikv/tikv/pull/8865#discussion_r510085200

Reply:
okay